stratum/hal/lib/tdi/tdi_sde_wrapper.cc

```diff
@@ -3336,302 +374,422 @@ ::util::Status TdiSdeWrapper::GetActionProfileGroups(
-    uint32 table_id, const TableDataInterface* table_data) {
+    uint32 table_id, absl::optional<uint32> meter_index, bool in_pps,
+    uint64 cir, uint64 cburst, uint64 pir, uint64 pburst) {
   ::absl::ReaderMutexLock l(&data_lock_);
   auto real_session = std::dynamic_pointer_cast<Session>(session);
   CHECK_RETURN_IF_FALSE(real_session);
-  auto real_table_data = dynamic_cast<const TableData*>(table_data);
-  CHECK_RETURN_IF_FALSE(real_table_data);
+
   const ::tdi::Table* table;
   RETURN_IF_TDI_ERROR(tdi_info_->tableFromIdGet(table_id, &table));
 
+  std::unique_ptr<::tdi::TableKey> table_key;
+  std::unique_ptr<::tdi::TableData> table_data;
+  RETURN_IF_TDI_ERROR(table->keyAllocate(&table_key));
+  RETURN_IF_TDI_ERROR(table->dataAllocate(&table_data));
+
+  // Meter data: $METER_SPEC_*
+  if (in_pps) {
+    RETURN_IF_ERROR(SetField(table_data.get(), kMeterCirPps, cir));
+    RETURN_IF_ERROR(
+        SetField(table_data.get(), kMeterCommitedBurstPackets, cburst));
+    RETURN_IF_ERROR(SetField(table_data.get(), kMeterPirPps, pir));
+    RETURN_IF_ERROR(SetField(table_data.get(), kMeterPeakBurstPackets, pburst));
+  } else {
+    RETURN_IF_ERROR(
+        SetField(table_data.get(), kMeterCirKbps, BytesPerSecondToKbits(cir)));
+    RETURN_IF_ERROR(SetField(table_data.get(), kMeterCommitedBurstKbits,
+                             BytesPerSecondToKbits(cburst)));
+    RETURN_IF_ERROR(
+        SetField(table_data.get(), kMeterPirKbps, BytesPerSecondToKbits(pir)));
+    RETURN_IF_ERROR(SetField(table_data.get(), kMeterPeakBurstKbits,
+                             BytesPerSecondToKbits(pburst)));
+  }
+
   const ::tdi::Device *device = nullptr;
   ::tdi::DevMgr::getInstance().deviceGet(dev_id, &device);
   std::unique_ptr<::tdi::Target> dev_tgt;
   device->createTarget(&dev_tgt);
 
   ::tdi::Flags *flags = new ::tdi::Flags(0);
-  RETURN_IF_TDI_ERROR(table->defaultEntrySet(
-      *real_session->tdi_session_, *dev_tgt,
-      *flags, *real_table_data->table_data_));
+  if (meter_index) {
+    // Single index target.
+    // Meter key: $METER_INDEX
+    RETURN_IF_ERROR(
+        SetFieldExact(table_key.get(), kMeterIndex, meter_index.value()));
+    RETURN_IF_TDI_ERROR(table->entryMod(
+        *real_session->tdi_session_, *dev_tgt, *flags, *table_key, *table_data));
+  } else {
+    // Wildcard write to all indices.
+    size_t table_size;
+    RETURN_IF_TDI_ERROR(table->sizeGet(*real_session->tdi_session_,
+                                       *dev_tgt, *flags, &table_size));
+    for (size_t i = 0; i < table_size; ++i) {
+      // Meter key: $METER_INDEX
+      RETURN_IF_ERROR(SetFieldExact(table_key.get(), kMeterIndex, i));
+      RETURN_IF_TDI_ERROR(table->entryMod(
+          *real_session->tdi_session_, *dev_tgt, *flags, *table_key, *table_data));
+    }
+  }
+
   return ::util::OkStatus();
 }
 
-::util::Status TdiSdeWrapper::ResetDefaultTableEntry(
+::util::Status TdiSdeWrapper::ReadIndirectMeters(
```

Comment:
Meters are also special table types like action profiles. They can also probably get their own files.

Reply:
True. Meters, Counters, Registers, and CloneSessions are all candidates for additional refactoring.